src/BloomExe/web/EnhancedImageServer.cs

```diff
@@ -1,51 +1,137 @@
-﻿// Copyright (c) 2014 SIL International
+﻿// Copyright (c) 2014-2015 SIL International
 // This software is licensed under the MIT License (http://opensource.org/licenses/MIT)
 using System;
 using System.Diagnostics;
-using System.Drawing.Text;
 using System.Globalization;
-using System.Linq;
-using System.Net;
-using Bloom.ImageProcessing;
 using System.IO;
-using L10NSharp;
+using System.Threading;
 using Microsoft.Win32;
+using L10NSharp;
 using Palaso.IO;
 using Bloom.Collection;
-
+using Bloom.ImageProcessing;
 
 namespace Bloom.web
 {
 	/// <summary>
 	/// A local http server that can serve (low-res) images plus other files.
 	/// </summary>
+	/// <remarks>geckofx makes concurrent requests of URLs which this class handles. This means
+	/// that the methods of this class get called on different threads, so it has to be
+	/// thread-safe.</remarks>
 	public class EnhancedImageServer: ImageServer
 	{
 		private FileSystemWatcher _sampleTextsWatcher;
 		private bool _sampleTextsChanged = true;
+		private const int RWLockTimeOut = 100;
 
 		public CollectionSettings CurrentCollectionSettings { get; set; }
 
 		public EnhancedImageServer(LowResImageCache cache): base(cache)
 		{
+			Lock = new ReaderWriterLock();
+		}
+
+		// We use three different locks to synchronize access to the methods of this class.
+		// This allows certain methods to run concurrently.
+
+		// used to synchronize access to properties. Use this lock when you want to allow
+		// multiple concurrent readers, but no one else when writing the property.
+		private ReaderWriterLock Lock { get; set; }
+		// used to synchronize access to I18N methods
+		private object I18NLock = new object();
+		// used to synchronize access to various methods
+		private object SyncObj = new object();
+
+		private string _currentPageContent;
+		private string _accordionContent;
+		private bool _authorMode;
+
+		public string CurrentPageContent
```

Comment:
It would however be possible that a thread reads CurrentPageContent for a certain page and AccordionContent for a different page, if some other thread is busy updating both. I'm not sure whether AccordionContent actually changes from page to page, but I think it is only updated in one place so it MAY be worth having that code get a single write lock, or (better) making a new method to update both at once in a single write-lock, and making the actual properties read-only.